8193800: TreeTableView selection changes on sorting

arapte opened the pull request:
Issue:
In TreeTableView, in case of Multiple selection mode, if nested items are selected, then TreeTableView does not retain/update the selection correctly when the tree items are permuted(either by `sort()` or by reordering using `setAll()`).

Cause:

1. For permutation, the current implementation uses `TreeModificationEvent` to update the selection.
2. The indices from these TreeModificationEvents are not reliable.
3. It uses the non public `TreeTablePosition` constructor to create intermediate `TreeItem` positions, this constructor results in another unexpected TreeModificationEvent while one for sorting is already being processed.
4. In case of sorting, there can be multiple intermediate TreeModificationEvents generated, and for each TreeModificationEvent, the selection gets updated and results in selection change events being generated.
5. Each time a TreeItem is expanded or collapsed, the selection must be shifted, but shifting is not necessary in case of permutation.
All these issues combine in wrong update of the selection.

Fix:

1. On each TreeModificationEvent for permutation, for updating the selection, use index of TreeItem from the TreeTableView but not from the TreeModificationEvent.
2. Added a new non public TreeTablePosition constructor, which is almost a copy constructor but accepts a different row.
3. In case of sorting, send out the set of selection change events only once after the sorting is over.
4. In case of setAll, send out the set of selection change events same as before.(setAll results in only one TreeModificationEvent, which effectively results in only one set of selection change events).
`shiftSelection()` should not be called in case of permutation i.e. call `if (shift != 0)`

Verification:
The change is very limited to updating of selection of TreeTableView items when the TreeItems are permuted, so the change should not cause any other failures.
Added unit tests which fail before and pass after the fix.
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8193800](https://bugs.openjdk.java.net/browse/JDK-8193800): TreeTableView selection changes on sorting


### Reviewers
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)
 * Ajit Ghaisas ([aghaisas](@aghaisas) - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/244/head:pull/244`
`$ git checkout pull/244`

kevinrushforth commented:
/reviewers 2

kevinrushforth commented:
@aghaisas can you also review this?

kevinrushforth commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java:
```diff
@@ -1804,6 +1804,11 @@ public int getVisibleLeafIndex(TreeTableColumn<S,?> column) {
     }
 
+    boolean sortingInProgress;
+    protected boolean isSortingInProgress() {
```
This is an implementation detail. By making it `protected` it becomes public API. You will need to make it package-scope instead

arapte replied:
Updated in next commit

kevinrushforth commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java:
```diff
@@ -1814,6 +1819,7 @@ public int getVisibleLeafIndex(TreeTableColumn<S,?> column) {
      * something external changes and a sort is required.
      */
     public void sort() {
+        sortingInProgress = true;
```
I presume that sorting cannot be called recursively?

arapte replied:
The `sort()` is called only once for a `TreeTableView`. This method does not perform the actual sorting but it is performed by the sort policy(`DEFAULT_SORT_POLICY`). 
`DEFAULT_SORT_POLICY` initiates the sorting by calling `rootItem.sort()`, which results in call to `TreeItem.doSort()`.
`TreeItem.doSort()` makes the final call to sort the children list, this method gets executed once for the root TreeItem and then for each expanded child TreeItem.

kevinrushforth commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java:
```diff
@@ -1804,6 +1804,11 @@ public int getVisibleLeafIndex(TreeTableColumn<S,?> column) {
         return visibleLeafColumns.get(column);
     }
 
+    boolean sortingInProgress;
```
Minor: the field itself can be private.

arapte replied:
Corrected in the next commit.

kevinrushforth commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTablePosition.java:
```diff
@@ -77,6 +77,17 @@ public TreeTablePosition(@NamedArg("treeTableView") TreeTableView<S> treeTableVi
 
+    // A call to 'treeTableView.getTreeItem(row)' may result in TreeModificationEvent being triggered.
+    // This causes issue by triggering a new TreeModificationEvent while one TreeModificationEvent
+    // is being handled currently.
```
The first three lines of added comments don't really belong here. I would just document the new constructor as a copy-like constructor (with a different row) and mention that it is used by the TreeTableView::sort method.

arapte replied:
Updated the doc in next commit, please take a look

arapte commented:
> The algorithm looks correct to me for sorting. How much regression testing have you done for cases where rows or columns are inserted?

I have tested with a small TreeTableView of 15 rows of 3 levels and 3 columns.
Would it be enough to test with TreeTableView of ~500 rows of 5 levels, 10 columns ?

kleopatra commented:
hmm .. TreeModificationEvent seems to have a different interpretation (than a list change) of _permutated_: its wasPermutated may return true even on a replace - that's why some tests are throwing a IllegalStateException before the fix. 

The other way round: do we really want to introduce a singularity in selection handling after replace? The other selectionModels for virtualized controls try to keep the selectedItem/Index only.

kleopatra commented:
ahh .. was wrong (note to self: run examples for all controls before commenting ;) - TableView also tries to keep all selected state on replace with the same items in changed sequence, ListView doesn't.

kleopatra commented:
need a break ;) 

Will come back later with an example comparing the behavior of multiple selection state across virtualized controls - preliminary results

Sort:
-  ListView/TableView keep selection (corrrectly) on all selectedItems, adjust indices as needed
- TreeView: ??
- TreeTableView: throws IllegalState before this fix, keeps selection (correctly) on all selectedItems, adjusts indices as needed

Replace with same items in different sequence
- ListView/TableView: keep only selectedItem, adjust selectedIndex as needed
- TreeView: keeps selectedIndices (?)
- TreeTableView: throws IllegalState before this fix, keeps selectedItems, adjusts selectedIndices as needed

kevinrushforth commented:
> I have tested with a small TreeTableView of 15 rows of 3 levels and 3 columns.
> Would it be enough to test with TreeTableView of ~500 rows of 5 levels, 10 columns ?
> 
> Update: I am testing with 7 level of nested rows, with 10, 9, 7, 6, 5, 4, 3 number of children in each level respectively. The fix works fine till level 3. But can observe issue with level 4 and further. Shall debug this more and update.

OK, thanks. In addition to making sure that insertion, deletion, and sorting work, are you also checking that the events being sent are as expected?

kleopatra commented:
Unrelated to this fix I see two pain points (which were not introduced here :)

#### A. replaced vs. permutated

These are types of change as decided by the list (aka: model) - treeModificationEvent (aka: view) must follow the lead (vs. spreading its own interpretation). Doing so is a bug, IMO.

#### B. treeTableView.sort changing internals of the selectionModel

that's a no-no-no-go, always. Instead let the selectionModel handle the permutated state correctly (didn't dig why it was deemed necessary to do so in sort - though it has the side-effect of leaving out selectionModels that are not of the expected type)

arapte commented:
> Update: I am testing with 7 level of nested rows, with 10, 9, 7, 6, 5, 4, 3 number of children in each level respectively. The fix works fine till level 3. But can observe issue with level 4 and further. Shall debug this more and update.

In the above case, The selection updates correctly but the received change event on SelectedItems after sort is incorrect. As of now it looks like the selection is updated correctly with this fix, but the event generated from sort() method is always incorrect. Looking in to fix the change event....

arapte commented:
> The other selectionModels for virtualized controls try to keep the selectedItem/Index only.

Keeping the selection after sort/permutation seems more correct from a user's perspective. I did not look into how other controls handle this. Looking at your observation seems like we need to fix this in other controls too, preferably each control in a separate issue :)

ebresie commented:
Do any of these variables(inProgress)/methods (sort) need to be made thread safe to avoid race conditions in some way?

arapte commented:
> In the above case, The selection updates correctly but the received change event on SelectedItems is incorrect. As of now it looks like the selection is updated correctly with this fix, but the event generated from sort() method is always incorrect. Looking in to fix the change event....

Please take a look at the updated changes. The `GenericAddRemoveChange` event generated from `TreeTableView.sort()` method does not result in correct selection change events. @kevinrushforth , Thanks for the guidance on this in offline discussion. The solution is to sort the tree of selected items upfront before generating the `GenericAddRemoveChange` event. Rest of the TreeItems can be left to be sorted lazily whenever accessed.

arapte commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java:
```diff
@@ -2540,68 +2565,40 @@ private void updateTreeEventListener(TreeItem<S> oldRoot, TreeItem<S> newRoot) {
+                            if (treeTableView.isSortingInProgress()) {
+                                startAtomic();
+                                selectedCellsMap.setAll(updatedSelection);
+                                stopAtomic();
```
This code block gets executed only as a result of sorting performed by `TreeTableView.sort()` method. The method `TreeTableView.sort()` makes call to the `SelectionModel.startAtomic()` and `SelectionModel.stopAtomic()`. Hence, Ideally it is not required to call those methods from here. But keeping these for precautions to safe guard against any future changes and better understanding when reading.

kevinrushforth commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java:
```diff
@@ -1809,6 +1810,11 @@ boolean isSortingInProgress() {
 
+    private boolean sortTreeOfSelectedItems = true;
+    boolean isSortTreeOfSelectedItems() {
+        return sortTreeOfSelectedItems;
```
This method / state attribute isn't really needed. The value is never set to anything other than `true`. I recommend removing it.

arapte replied:
Corrected in the next commit..

kevinrushforth commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewTest.java:
```diff
@@ -374,83 +374,133 @@ private String debug() {
+            treeRoot.getChildren().add(ti);
+
+            for (int j = 0; j < FIRST_LEVEL_COUNT - 1; j++) {
+                TreeItem<String> tj = new TreeItem<>("" + i + j);
```
The tree item strings will not be unique. This won't affect whether the test passes or fails, since `TreeItem` does not override `equals`, but it might be better if the strings were unique (in case there ever was an error it would be easier to understand).

arapte replied:
Hi Kevin, I verified the the tree item strings again. They seem to be unique. A total of 8800 of tree items get created each with unique string. Could you please recheck.

kevinrushforth replied:
Confirmed. This was my mistake.

kevinrushforth commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewTest.java:
```diff
@@ -374,83 +374,133 @@ private String debug() {
 
         sm.selectedIndexProperty().addListener(ov -> {
             countSelectedIndexChangeEvent++;
+            assertEquals(selectedItemBefore, treeTableView.getTreeItem(sm.getSelectedIndex()));
```
If this assertion ever fails, I don't think it will cause a test failure, since the event handling code will swallow the exception. The same is true of the other listeners. I don't know if it would be possible to use the UncaughtExceptionHandler as is done in other tests -- see [JDK-8244531](https://bugs.openjdk.java.net/browse/JDK-8244531) -- but that might be worth exploring. Another possibility is to wrap all the listeners in a try/catch and keep a list of `Throwable`s that are caught.

arapte replied:
Hi Kevin, I tested these listeners by adding assert that always fail. It shows the failures correctly. Also in this file there are other tests which assert inside a listener. I confirmed that those tests also show failures correctly. I could not find out what makes it work though.

kevinrushforth replied:
I can confirm that the test fails if an assertion hits. The reason is that this ChangeListener throws any exception to the code that causes the change, so it will propagate to the test thread.

kevinrushforth commented:
Pending a second reviewer.

@aghaisas or @kleopatra can one of you review it?

aghaisas commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTablePosition.java:
```diff
@@ -77,6 +77,16 @@ public TreeTablePosition(@NamedArg("treeTableView") TreeTableView<S> treeTableVi
         nonFixedColumnIndex = treeTableView == null || tableColumn == null ? -1 : treeTableView.getVisibleLeafIndex(tableColumn);
     }
 
```
Suggest to add 
    // Not public API
before this description.

(Similar to the constructor above this newly added constructor)

arapte replied:
Corrected in the next commit.

aghaisas commented on modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java:
```diff
@@ -1848,6 +1856,16 @@ public void sort() {
         if (sortPolicy == null) return;
```
Do we need to set sortingInProgress to false before returning from here?

kevinrushforth replied:
Yes, I think so. Good catch.

arapte replied:
Corrected in the next commit.

kleopatra commented:
confirmed test failures before and passing after the fix - impressive :)

While digging a bit (mainly around my [earlier concerns](https://github.com/openjdk/jfx/pull/244#issuecomment-638702570) - which still hold but could be discussed in a follow-up issue :) I came up with a NPE thrown by the newly added code block ins sort (the one walking up the parent hierarchy). 

    @Test
    public void testNPEOnSortAfterSetAll() {
        // stand-alone setup
        TreeTableView<String> treeTableView = new TreeTableView<>();
        TreeTableColumn<String, String> col = new TreeTableColumn<String, String>("column");
        col.setSortType(ASCENDING);
        col.setCellValueFactory(param -> new ReadOnlyObjectWrapper<String>(param.getValue().getValue()));
        treeTableView.getColumns().add(col);

        TreeItem<String> root = new TreeItem<String>("root");
        root.setExpanded(true);
        root.getChildren().addAll(
                new TreeItem("Apple"),
                new TreeItem("Orange"),
                new TreeItem("Banana"));

        treeTableView.setRoot(root);
        // add expanded children
        root.getChildren().forEach(child -> {
            child.setExpanded(true);
            String value = child.getValue();
            for (int i = 1; i <= 3; i++) {
                child.getChildren().add(new TreeItem<>(value + i));
            }    
        });
        assertEquals("sanity", 13, treeTableView.getExpandedItemCount());
        TreeTableViewSelectionModel<String> selectionModel = treeTableView.getSelectionModel();
        selectionModel.setSelectionMode(SelectionMode.MULTIPLE);
        selectionModel.selectIndices(2, 5, 8, 10);
        assertEquals(10, selectionModel.getSelectedIndex());
        TreeItem<String> lastRootChild = root.getChildren().get(2);
        assertEquals("sanity: selectedItem child of last", 
                lastRootChild, selectionModel.getSelectedItem().getParent());
        // replace children of last root child
        List<TreeItem<String>> childrenOfLastRootChild = new ArrayList<>(lastRootChild.getChildren());
        childrenOfLastRootChild.remove(0);
        lastRootChild.getChildren().setAll(childrenOfLastRootChild);
        treeTableView.sort();
    }

(before the fix it throws an IOOB, plus the behavior completely rotten in a visual test)

Did nothing to go to the bottom of this - looks like somehow the state of the selection is (still?) broken after setAll. 

Which might bring us back to the replace != permutation (can't resist to try :)  actually there's no reason to special case a replace with all same items against a replace with only some same items (doing so introduces a discontinuity in behavior) - we should either try to keep selected items or not.

arapte commented:
> While digging a bit (mainly around my [earlier concerns](https://github.com/openjdk/jfx/pull/244#issuecomment-638702570) - which still hold but could be discussed in a follow-up issue :)

Yes, We need to address these issues. How should we treat permutation vs replace, which may involve answering few more small questions. Like, What should be the selected item when we remove the current selected item? What should the contents of an event be in such cases? 

Regarding the scenario in your test, In `TreeTableView`, when `setAll()` is called by providing the exact same items in a different order, then it is considered as permutation. and a `wasPermutated` `TreeModificationEvent` gets generated. But if even a single `TreeItem` is removed and `setAll(` with the remaining n-1 TreeItems`)` is called, then it is considered as adding new items. (May be it should be generate as `wasRemoved`). And similar issue can be observed when a `TreeItem` is added.

In the test that you provided, one `TreeItem` is removed before doing `setAll()`. So this results in a `wasAdded` `TreeModificationEvent `and does not enter the `wasPermutated` code block. This fix only changes the `wasPermutated` event. So the behavior is still buggy more or less similar to before this change. There are multiple such issues with updating the selection. I have grouped few already reported issues under [JDK-8248217](https://bugs.openjdk.java.net/browse/JDK-8248217).

> I came up with a NPE thrown by the newly added code block ins sort (the one walking up the parent hierarchy).

I have added a null check for now, with a reasoning comment. We will have to remove that check once we fix the other issues.
Thanks for the providing the test :). code really helps to understand. I have added this scenario to test along with few others. The test are ignored using [JDK-8248217](https://bugs.openjdk.java.net/browse/JDK-8248217).
Please take a look.

kevinrushforth commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewTest.java:
```diff
@@ -385,32 +389,94 @@ private String debug() {
     }
 
-    @Test public void testSelectionUpdatesCorrectlyAfterChildSetAll() {
+    @Ignore("JDK-8248217")
```
Question: did you mean to point to the umbrella Task or is there a more specific bug that you could point to (I think referring to the Task is fine if you aren't sure). This applies to all of the newly added `@Ignore`d tests.

arapte replied:
These ignored tests are related to removing and adding a TreeItem.
I have corrected the id to use JDK-8193442 for the tests that remove an item and created a new JBS issue JDK-8248389, and used it for the test of adding an item .

kevinrushforth commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewTest.java:
```diff
@@ -385,32 +389,94 @@ private String debug() {
+
+    @Ignore("JDK-8248217")
+    @Test public void testSelectionUpdatesCorrectlyAfterChildRemoveOneAndSetAll() {
+        TreeTableColumn<String, String> col = setupForPermutationTest();
```
Minor: `col` is unused.

arapte replied:
Corrected in the next commit

kleopatra commented:
> 
> 
> The changes to the fix itself look fine. I have a question and minor comment on the test changes.
> 
> As for the questions raised by @kleopatra as long as you plan to address them in the follow-up issue(s), that seems fine, since you aren't regressing any behavior.

agreed - plus it's not only not regressing, but considerably improving the behavior :)

arapte commented:
/integrate
